### PR TITLE
Improve performance of date_histogram sometimes

### LINKED ISSUE #75542: Performance regression on aggs when top level query is single-node bool
@DJRickyB tracked down a performance regression on a sequence of aggs folks use to visualize logs. The rally configuration looks like this:

```
                "query": {
                  "bool": {
                    "must": [],
                    "filter": [
                      {
                        "match_all": {}
                      },
                      {
                        "range": {
                          "@timestamp": {
                            "gte": "2020-12-14T11:02:06.523Z",
                            "lte": "2020-12-15T11:02:06.523Z",
                            "format": "strict_date_optional_time"
                          }
                        }
                      }
                    ],
                    "should": [],
                    "must_not": []
                  }
                },
                "aggs": {
                  "ad64cabc-0bcf-4c39-ab6c-cedd90396a23": {
                    "terms": {
                      "field": "host.name",
                      "order": {
                        "_count": "desc"
                      },
                      "size": 5
                    },
                    "aggs": {
                      "97997d07-3f9f-424f-bb6d-9c698013e64f": {
                        "date_histogram": {
                          "field": "@timestamp",
                          "fixed_interval": "30m",
                          "time_zone": "UTC"
                        }
                      }
                    }
                  }
                }
```

It looks like #74260 slowed it down significantly. I'd test things like this locally when implementing #74260 and didn't see a regression, but the performance testing infrastructure is a lot better at tracking this sort of thing. I'll pick this up soon-ish. Probably not for a few days, but I would like to at least understand it more closely.

image

We aren't going to revert #74260 because it is more important not to make complex queries much, much worse. But it is a real shame to have hit the 90% case as hard as we did. I don't imagine we'll have a fix for this regression in 7.14 either, sadly.

### LINKED ISSUE #82384: Use filter by filter if top level query is a bool wrapping a match all
Right now filter aggregations run in "compatible" mode if the top level query is `bool: {filter: { match_all: {}}}}` - but this really just just `match_all` in disguise. We should just run in filter-by-filter mode here.

### PR DESCRIPTION
**Pull request by nik9000**
When the top level query is a boolean like:
```
                "query": {
                  "bool": {
                    "must": [],
                    "filter": [
                      {
                        "match_all": {}
                      },
                      {
                        "range": {
                          "@timestamp": {
                            "gte": "2020-12-14T11:02:06.523Z",
                            "lte": "2020-12-15T11:02:06.523Z",
                            "format": "strict_date_optional_time"
                          }
                        }
                      }
                    ],
                    "should": [],
                    "must_not": []
                  }
                },
```

Our filter-by-filter optimization mistakenly didn't kick in even when
it'd be faster. Convinces it to kick in which should speed things up.

Closes #75542
Closes #82384

### CODE REVIEW

**Comment by nik9000** [comment]
I'm going to run some performance tests on this to confirm it is doing its job.

**Comment by nik9000** [comment]
run elasticsearch-ci/bwc

**Comment by nik9000** [comment]
I've added a few other logging statements that were useful in debugging this.

**Review by not-napoleon — APPROVED** [review]
LGTM

File: docs/changelog/83751.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+pr: 83751
+summary: Improve performance of `date_histogram` sometimes
```
[the remark this answers is not in the document]
Comment (nik9000) [reply docs/changelog/83751.yaml:2]: I suppose I can update the description here.

**Comment by nik9000** [comment]
@elasticmachine update branch

**Comment by jpountz** [comment]
@nik9000 I'm wondering if this change is a candidate for backporting to 7.17, as it could be seen as a performance regression by some users. I don't have a strong opinion on the question, curious what you think.

**Comment by inqueue** [comment]
+1 for backporting to 7.17 if you think it is reasonable.

**Comment by nik9000** [comment]
I think only 7.13 would have had the "best" performance in this case. 7.12 didn't use filter-by-filter here and 7.14 turned it off.

I can backport, but I don't know that it's really worth doing without it's companion #83902. The case we were hitting in rally really wanted both fixes in order to really hit. And because they were performance work I didn't backport. I certainly can. They are contained enough.

**Pull request merged**